content/general/concepts/what-is-the-prototype-pattern/what-is-the-prototype-pattern.md

```diff
@@ -0,0 +1,266 @@
+---
+Title: 'What is the Prototype Pattern?'
+Description: 'The prototype pattern adds cloning functionality to an objects class.'
+Subjects:
+- 'Computer Science'
+- 'Interview Prep' 
+Tags:
+- 'Classes'
+- 'Conceptual'
+- 'Constructors'
+- 'Objects'
+CatalogContent:
+- 'learn-java'
+- 'paths/computer-science'
+---
+
+The _prototype pattern_ adds cloning functionality to an object's class. After the instantiation an initial object, the prototype removes the need for the `new` keyword in creating subsequent objects at run time. New objects are created using the property values of a pre-existing object.
+
+There are two types of copying associated with the prototype pattern. They are:
+
+- Shallow copy: Copies the immediate property values.
+- Deep copy: Copies the immediate values, plus any referenced object.
+
+## Example: Shallow Copy vs Deep Copy
+
+To illustrate the difference between the shallow and deep copy, a model has been provided below, consisting of two classes `TvSeries` and `Episode`. Cloning functionality has been added to `TvSeries` by implementing the `Cloneable` interface.
+
+```java
+public class TvSeries implements Cloneable {
+  private final String name;
+  private final int seriesNo;
+  // Notice one or more properties are a complex type and mutable
+  private final List<Episode> episodes;
+
+  public TvSeries(String name, int seriesNo, List<Episode> episodes) {
+    this.name = name;
+    this.seriesNo = seriesNo;
+    this.episodes = episodes;
+  }
+
+  // Shallow copy method
+  @Override
+  public TvSeries clone() throws CloneNotSupportedException {
+    // Cast to TvSeries as super returns Object
+    return (TvSeries) super.clone();
+  }
+
+  // Deep copy
+  public TvSeries deepCopy() {
+    // New Episode objects are created during the copy
+    return new TvSeries(this.name, this.seriesNo, this.episodes.stream()
+                                                               .map(episode -> new Episode(episode.getName(), episode.getEpNo()))
+                                                               .collect(Collectors.toList()));
+  }
```

Comment:
The code here looks a bit awkward

Reply:
In addition, it's important to keep this user-friendly! I feel that these code snippets are quite lengthy and difficult to follow. Is it possible to shorten these examples and still get the original point across? Every word counts!

Reply:
In addition, it's important to keep this user-friendly! I feel that these code snippets are quite lengthy and difficult to follow. Is it possible to shorten these examples and still get the original point across? Every word counts!

Thanks for your contributions 🥇

Reply:
Hi,

Yes this is starting to become a little bit of an issue for me (more so in the factory.) When I first wrote this code e.g. it returned the user's last 10 watched shows regardless of whether they were films or tvshows. I then reduced it as much as I could. The issue is that these design patterns have quite a lot of moving parts to explain for a short entry definition, and I wanted to try and find an e.g. that might make some sense in the real world rather than explaining in the abstract (hence the mock repository.) Understanding how to de couple responsibilities at the class level is really important, and design patterns seem to give us a common language / way to do this. Unfortunately this leads to multiple classes. (<- I'm not totally sure this (real world thing) is the direction I should be going in. Explaining my thoughts and trying to impart conceptual ideas is still very new to me.)

With regards to the code looking a bit awkward, I would almost always prefer functional over classical (just because it can be done in one line) but sure. This copy can be done using a for each as well. No worries Ill change that after work today.

I do want this to help others and you comments are incredibly valuable to me. As my code e.g. makes sense to me by the time I write the entry, I'm never sure how clear my explanation to others are. I do want these entries to be as useful as possible.

Thank you! I'll see what I can do :)

Reply:
Thanks for putting in the effort for a workaround!

Creating real-world scenarios is really helpful and I don't think that aspect needs to be changed. I get what you mean though about there being so many moving parts and the need to incorporate as much code content as possible. One way you can go about seeing what to keep or remove is by deciding what you can fully unpack from a code snippet and work around those fine details. Every line of code should be easily dissectable through the explanation and actual snippet.

💪

Reply:
Hi,

Please see my code change and Ive added more explanation. To be honest I feel I am over explaining, but happy to be guided on this.

I have also left in the comments (not sure if this is useful or not.)

I wont mark as resolved yet, as want to hear your thoughts.

Thanks again

Reply:
I think the explanations are helpful! Really appreciate your responsiveness through this process. I'm just going to make some structural edits, double-check all the code, and then approve of this PR.

Thanks @Bernard-McGeever !